packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -1299,13 +1299,86 @@ class RenderFractionalTranslation extends RenderProxyBox {
   }
 }
 
+/// The interface used by [CustomPaint] (in the widgets library) and
+/// [RenderCustomPaint] (in the rendering library).
+///
+/// To implement a custom painter, subclass this interface to define your custom
+/// paint delegate.
+///
+/// The [paint] method is called whenever the custom object needs to be
+/// repainted. This is the primary method that must be implemented.
+///
+/// The [shouldRepaint] method must also be implemented. It is called when a new
+/// instance of a delegate class is provided, to check if the new instance
+/// actually represents different information.
+///
+/// The [hitTest] method can optionally be implemented also.
 abstract class CustomPainter {
   /// Abstract const constructor. This constructor enables subclasses to provide
   /// const constructors so that they can be used in const expressions.
   const CustomPainter();
 
+  /// Called whenever the object needs to paint. The given [Canvas] has its
+  /// coordinate space configured such that the origin is at the top left of the
+  /// box. The area of the box is the size of the [size] argument.
+  ///
+  /// Paint operations should remain inside the given area. Graphical operations
+  /// outside the bounds may be silently ignored, clipped, or not clipped; no
+  /// guarantees are made.
+  ///
+  /// Implementations should be wary of correctly pairing any calls to
+  /// [Canvas.save]/[Canvas.saveLayer] and [Canvas.restore], otherwise all
+  /// subsequent painting on this canvas may be affected, with potentially
+  /// hilarious but confusing results.
```

Comment:
s/"with potentially hilarious but confusing results"/"with unexpected results"/

Since hilarity is rather subjective, let's keep this plain :)